From 4ac5af8d5dbceaf4dc6c5e4a9729ce86bbb9efbd Mon Sep 17 00:00:00 2001
From: mdxd44 <mdxd44@ely.su>
Date: Sat, 19 Jun 2021 09:51:26 +0900
Subject: [PATCH] Update gradle


diff --git a/build.gradle b/build.gradle
index a322f1c2..ad8d0d75 100644
--- a/build.gradle
+++ b/build.gradle
@@ -3,19 +3,19 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.github.jengelman.gradle.plugins:shadow:5.0.0'
+        classpath 'com.github.jengelman.gradle.plugins:shadow:6.1.0'
     }
 }
 
 plugins {
     id 'java'
-    id "com.github.spotbugs" version "4.2.4" apply false
-    id 'org.cadixdev.licenser' version '0.5.1' apply false
+    id "com.github.spotbugs" version "4.7.1" apply false
+    id 'org.cadixdev.licenser' version '0.6.1' apply false
 }
 
 allprojects {
     apply plugin: "com.github.spotbugs"
-    // ElytraProxy - Fix UTF-8 for build on Windows
+    // ElytraProxy: Fix UTF-8 for build on Windows
     compileJava.options.encoding = 'UTF-8'
 
     group 'com.velocitypowered'
@@ -71,11 +71,11 @@ allprojects {
         }
     }
 
-    // Start ElytraProxy - Fix UTF-8 for build on Windows
+    // ElytraProxy Start: Fix UTF-8 for build on Windows
     tasks.withType(JavaCompile) {
         options.encoding = 'UTF-8'
     }
-    // End ElytraProxy
+    // ElytraProxy End
 
     tasks.withType(com.github.spotbugs.snom.SpotBugsTask) {
         reports {
diff --git a/gradle/wrapper/gradle-wrapper.properties b/gradle/wrapper/gradle-wrapper.properties
index 71796f88..4e8b1495 100644
--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.6.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.1-all.zip
diff --git a/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java b/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
index d9358c9e..b6c1cb07 100644
--- a/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java
@@ -184,7 +184,8 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
       // ElytraProxy
       implName = MoreObjects.firstNonNull(pkg.getImplementationTitle(), "ElytraProxy");
       implVersion = MoreObjects.firstNonNull(pkg.getImplementationVersion(), "<unknown>");
-      implVendor = MoreObjects.firstNonNull(pkg.getImplementationVendor(), "Velocity Contributors, ElytraProxy Contributors");
+      implVendor = MoreObjects.firstNonNull(pkg.getImplementationVendor(), "Velocity Contributors,"
+          + " ElytraProxy Contributors");
     } else {
       implName = "ElytraProxy";
       implVersion = "<unknown>";
@@ -453,8 +454,9 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
     }
 
     commandManager.setAnnounceProxyCommands(newConfiguration.isAnnounceProxyCommands());
+    // BotFilter backport: global rate limiter
     ipAttemptLimiter = Ratelimiters.createWithMilliseconds(newConfiguration.getLoginRatelimit(),
-        newConfiguration.getLoginRatelimitAttempts()); // BotFilter backport: global rate limiter
+        newConfiguration.getLoginRatelimitAttempts());
     this.configuration = newConfiguration;
     eventManager.fireAndForget(new ProxyReloadEvent());
     return true;
-- 
2.24.1.windows.2

